### PR TITLE
feat(material-experimental): add test harness for mat-tooltip

### PR DESCRIPTION
**Pull request by crisbeto**
Sets up a test harness for the current Material tooltip.

### CODE REVIEW

File: src/material-experimental/mdc-tooltip/harness/tooltip-harness.spec.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  it('should be able to open a tooltip', async () => {
+    const tooltip = await loader.getHarness(harness.with({id: 'one'}));
+    expect(await tooltip.isOpen()).toBe(false);
+    await tooltip.open();
+    expect(await tooltip.isOpen()).toBe(true);
+  });
+
+  // TODO(crisbeto): this fails because we have a setTimeout inside the tooltip which
+  // can't be flushed. Talk to Miles about how we should approach this in harnesses.
+  // it('should be able to close a tooltip', async () => {
```
Comment (crisbeto) [inline src/material-experimental/mdc-tooltip/harness/tooltip-harness.spec.ts]: @mmalerba In the tooltip's `hide` method we do something like `setTimeout(/* some close logic */, 0)` which I wasn't able to flush during the unit tests. I tried `await`ing a `setTimeout` as well, but it didn't help. I could probably get it to work if I use `fakeAsync` with a `tick` for the test, but that means that whoever is consuming the harness would have to know to use it as well. How should we handle this case? I'm guessing that we don't to lock ourselves into using `fakeAsync` or `async` around some tests.

File: src/material-experimental/mdc-tooltip/harness/tooltip-harness-filters.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+export type TooltipHarnessFilters = {
+  id?: string;
```
Comment (jelbourn) [inline src/material-experimental/mdc-tooltip/harness/tooltip-harness-filters.ts]: `triggerId`?

File: src/material-experimental/mdc-tooltip/harness/tooltip-harness.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+  /** Gets the overlay that contains the tooltip content. */
+  async getTooltipElement(): Promise<TestElement> {
+    return this._panel();
+  }
+
+  /** Gets an attribute value from the tooltip trigger. */
+  async getAttribute(attributeName: string): Promise<string|null> {
+    return (await this.host()).getAttribute(attributeName);
+  }
```
Comment (jelbourn) [inline src/material-experimental/mdc-tooltip/harness/tooltip-harness.ts]: I think we should just let people do `host().getAttribute` themselves if they really need to get this level of flexibility

File: src/material-experimental/mdc-tooltip/harness/tooltip-harness.spec.ts (hunk 1 above)
Comment (jelbourn) [reply src/material-experimental/mdc-tooltip/harness/tooltip-harness.spec.ts]: When you say `await`ing a `setTimeout`, what did that code look like?

Comment (crisbeto) [reply src/material-experimental/mdc-tooltip/harness/tooltip-harness.spec.ts]: I had something like this:

```ts
function wait(time: number) {
    return new Promise(resolve => {
        setTimeout(resolve, time);
    });
}
```

Then I had `await wait(20000)` right before the assertions, but the timer still hadn't fired.

**Comment by crisbeto** [comment]
Closing since this too out of date now. I'll take whatever I can from here and resubmit.

**Pull request closed without merging**